lib/promscrape: flaky test TestClientProxyReadOk

### Describe the bug

The [`TestClientProxyReadOk` test at `lib/promscrape`](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/6f61e9d49d3604a7960cd637666853c34a93debd/lib/promscrape/client_test.go#L116) sometimes fails with the following error message:

```
--- FAIL: TestClientProxyReadOk (0.03s)
    client_test.go:178: unexpected error at ReadData: cannot perform request to "https://127.0.0.1:46509": Get "https://127.0.0.1:46509": EOF
```

### To Reproduce

Run `make test-full` command from VictoriaMetrics repository root in a loop until the test `TestClientProxyReadOk` fails. It looks like it fails more frequently when `make test-full` command is executed in `cluster` branch.

### Version

The latest commits in `master` and `cluster` branches.

### Logs

_No response_

### Screenshots

_No response_

### Used command-line flags

_No response_

### Additional information

The `TestClientProxyReadOk` test was added by @f41gh7 in the commit 9feee15493343ef385ae47b3238cb15450f67a86 ( https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6783 ). There was an unsuccessful attempt to fix the flaky test by @hagen1778 in the commit f586082520dda509d5398a777c2fce83ed39e214 ( see https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6899 ).

@f41gh7 , @rtm0 , could you look closer at this test and fix it?

--- a/lib/promscrape/client_test.go
+++ b/lib/promscrape/client_test.go
@@ -1,6 +1,21 @@
 package promscrape
 
-/*
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/bytesutil"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/promauth"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/proxy"
+)
+
 func copyHeader(dst, src http.Header) {
 	for k, vv := range src {
 		for _, v := range vv {
@@ -9,13 +24,18 @@ func copyHeader(dst, src http.Header) {
 	}
 }
 
+type connReadWriteCloser struct {
+	io.Reader
+	io.WriteCloser
+}
+
 func proxyTunnel(w http.ResponseWriter, r *http.Request) {
 	transfer := func(src io.ReadCloser, dst io.WriteCloser) {
 		defer dst.Close()
 		defer src.Close()
 		io.Copy(dst, src) //nolint
 	}
-	destConn, err := net.DialTimeout("tcp", r.Host, 10*time.Second)
+	server, err := net.DialTimeout("tcp", r.Host, 10*time.Second)
 	if err != nil {
 		http.Error(w, err.Error(), http.StatusServiceUnavailable)
 		return
@@ -26,12 +46,16 @@ func proxyTunnel(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, "Hijacking not supported", http.StatusInternalServerError)
 		return
 	}
-	clientConn, _, err := hijacker.Hijack()
+	clientConn, clientBuf, err := hijacker.Hijack()
 	if err != nil {
 		http.Error(w, err.Error(), http.StatusServiceUnavailable)
 	}
-	go transfer(clientConn, destConn)
-	transfer(destConn, clientConn)
+	// For hijacked connections, one has to read from the connection buffer, but
+	// still write directly to the connection.
+	client := &connReadWriteCloser{clientBuf, clientConn}
+
+	go transfer(client, server)
+	transfer(server, client)
 }
 
 type testProxyServer struct {
@@ -133,14 +157,7 @@ func TestClientProxyReadOk(t *testing.T) {
 		}
 
 		var bb bytesutil.ByteBuffer
-		err = c.ReadData(&bb)
-		if errors.Is(err, io.EOF) {
-			bb.Reset()
-			// EOF could occur in slow envs, like CI
-			err = c.ReadData(&bb)
-		}
-
-		if err != nil {
+		if err = c.ReadData(&bb); err != nil {
 			t.Fatalf("unexpected error at ReadData: %s", err)
 		}
 		got, err := io.ReadAll(bb.NewReader())
@@ -183,4 +200,3 @@ func TestClientProxyReadOk(t *testing.T) {
 	// backend tls and proxy auth
 	f(true, false, nil, &promauth.BasicAuthConfig{Username: "proxy-test", Password: promauth.NewSecret("1234")})
 }
-*/

